### ISSUE TITLE
Treesitter performance (tracking issue)

### ISSUE THREAD

**Issue by lewis6991** [neovim/neovim#22426]
### Problem

Our treesitter implementation could be more efficient to overcome many known issues with performance. Specifically around large files.

### Steps to reproduce

dd

### Expected behavior

dd

### Neovim version (nvim -v)

dd

### Vim (not Nvim) behaves the same?

dd

### Operating system/version

dd

### Terminal name/version

dd

### $TERM environment variable

dd

### Installation

dd

**Comment by clason** [comment]
Another issue is incremental querying, which may help with expensive injection predicates as in https://github.com/nvim-treesitter/nvim-treesitter/issues/2996?

**Comment by lewis6991** [comment]
Yes it is. This will be partially resolved by just running the queries less in #22394. I haven't been able to triage anything specific about why the queries are slow.

**Comment by clason** [comment]
"Bad" files for performance testing: https://raw.githubusercontent.com/torvalds/linux/master/drivers/gpu/drm/amd/include/asic_reg/bif/bif_5_1_sh_mask.h

(cpp header consisting of 33k injections of cpp int preprocessor macros)

**Comment by clason** [comment]
Idea from chat: Highlighter should not parse off-screen injections (should help with those files).

**Comment by riskygit** [comment]
Given this project has a flake.nix I figure someone might know... is it possible to only remove "comment" while using `withAllGrammars` (or achieving "all other grammars" in some way)? Comment cripples performance so immensely hideously.

**Comment by lewis6991** [comment]
#24647 is now merged which should improve the performance of large files with lots of injections.

**Comment by MKrbm** [comment]
Hey gusy! I'm really appreciate with this plugin and your efforts on improving it!

I'd like to ask, in this issue, if you are dealing with the problem of treesitter starting to perform dramatically poorly for large file, especially when turning-on highlight and indent .

Thank you very much!

**Comment by MKrbm** [comment]
Btw, what does it mean by "injection" here? I thought it just talks about language injection within markdown files or something.

**Comment by tomtomjhj** [comment]
>  Partially run the query for rendered lines.

This would require separate queries for non-combined injection and combined injection (or some other significant API changes). Combined injection query should be run on the entire source because the invisble part of the region affects the visible part, just like the root language. So the partial query execution (using `start` and `stop` params of `iter_matches`) should be applied only to non-combined injections. However, currently there is no way to iterate only the matches that have some specific metadata. The simpliest solution I can think of is to have two separate query files for combined and non-combined ones.

**Comment by lewis6991** [comment]
Non-combined is the norm, so if combined is detected such optimisations should just be disabled.

I think adding additional states could over complicate things for not a lot of return.

**Comment by lewis6991** [comment]
Non-combined is the norm, so if combined is detected such optimisations should just be disabled.

I think adding additional states could over complicate things for not a lot of return.

**Comment by tomtomjhj** [comment]
> Partially run the query for rendered lines.

Another issue with this is that it's difficult to make injected region parsing incremental. This requires some sort of persistent ID for each region and the corresponding tree so that the old tree can be used when parsing the corresponding region. The current implementation of LanguageTree simply uses the index of the region in the whole list of regions captured by the last execution of the injection query, which is frequently invalidated. Unfortunately, treesitter doesn't seem to provide such a functionality. Using the equality of the region value could be an option, but that looks inefficient.

Even if there were some ID mechanism that allows incremental parsing of injected regions, I think it would be still inefficient because all those regions should be adjusted after each edit. So the parser would need to discard trees (e.g., with LRU cache) to balance the cost of maintaining the existing trees and parsing the new regions from scratch.

**Comment by lewis6991** [comment]
Helix apparently already does this so I suggest you look at there implementation.
